PR: https://github.com/odoo/odoo/pull/195611

From: c54e859885d6cbde799878c33078c4729d816e3f
From: alan-odoo
Date: 2025-08-25 10:20:52

Breaking data model changes scores: del:10 + add:6, change matches:
-    question_ids = fields.One2many(
-    general_question_ids = fields.One2many('event.question', 'event_id', 'General Questions',
-    specific_question_ids = fields.One2many('event.question', 'event_id', 'Specific Questions',
+    question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
+    general_question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
+    specific_question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
-    event_type_id = fields.Many2one('event.type', 'Event Type', ondelete='cascade', index='btree_not_null')
-    event_id = fields.Many2one('event.event', 'Event', ondelete='cascade', index='btree_not_null')
+    event_type_ids = fields.Many2many('event.type', string='Event Types', copy=False)
+    event_ids = fields.Many2many('event.event', string='Events', copy=False)
-    question_ids = fields.One2many(
+    question_ids = fields.Many2many(

Total Changes: 96

[IMP] event,*: make event questions m2m

* = pos_event, test_event_full, website_event

This commit allows to create questions that can be shared by events and also by
event templates. Therefore, if a question has to be used by several events or
event templates, there is no need to duplicate it. From now on, if a question
is modified, the change will be applied to both the events and the event types
to which it is linked.

As questions must be reusable between events and event templates, the fields
EventEvent.question_id, EventQuestion.event_id, EventType.question_ids,
EventQuestion.event_type_id have become many2many fields.

When modifying questions from the events and event templates, a warning
message is displayed indicating that the question is used by several events
or event templates and will be changed for all of them.

Previously, default questions were created for each events. It's not longer
the case. Default questions are created in db once and for all and are
reusable between events and event_templates.

Tests have been created to check if questions are correctly updated when the
event's template is changed and to check if new events are linked to the
default questions.

Upgrade PR: https://github.com/odoo/upgrade/pull/7153

Task-4247015

Part-of: odoo/odoo#195611
Related: odoo/upgrade#7153
Signed-off-by: Thibault Delavallee (tde) <tde@openerp.com>

================================= pseudo patch: =================================

--- a/addons/event/models/event_event.py
+++ b/addons/event/models/event_event.py
@@ -194,13 +194,12 @@ class EventEvent(models.Model):
         compute='_compute_ticket_instructions', store=True, readonly=False,
         help="This information will be printed on your tickets.")
     # questions
-    question_ids = fields.One2many(
-        'event.question', 'event_id', 'Questions', copy=True,
-        compute='_compute_question_ids', readonly=False, store=True, precompute=True)
-    general_question_ids = fields.One2many('event.question', 'event_id', 'General Questions',
-                                           domain=[('once_per_order', '=', True)])
-    specific_question_ids = fields.One2many('event.question', 'event_id', 'Specific Questions',
-                                            domain=[('once_per_order', '=', False)])
+    question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
+        string='Questions', compute='_compute_question_ids', readonly=False, store=True, precompute=True)
+    general_question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
+        string='General Questions', domain=[('once_per_order', '=', True)])
+    specific_question_ids = fields.Many2many('event.question', 'event_event_event_question_rel',
+        string='Specific Questions', domain=[('once_per_order', '=', False)])
 
     def _compute_use_barcode(self):
         use_barcode = self.env['ir.config_parameter'].sudo().get_param('event.use_event_barcode') == 'True'
@@ -220,18 +219,18 @@ class EventEvent(models.Model):
 
         When synchronizing questions:
 
-          * lines with no registered answers are removed;
+          * lines with no registered answers for the event are removed;
           * type lines are added;
         """
-        if self._origin.question_ids:
-            # lines to keep: those with already given answers
-            questions_tokeep_ids = self.env['event.registration.answer'].search(
-                [('question_id', 'in', self._origin.question_ids.ids)]
-            ).question_id.ids
-        else:
-            questions_tokeep_ids = []
         for event in self:
-            if not event.event_type_id and not event.question_ids:
+            questions_tokeep_ids = []
+            if self._origin.question_ids:
+                # Keep questions with attendee answers for the event.
+                questions_tokeep_ids.extend(
+                    (event.registration_ids.registration_answer_ids.question_id & self._origin.question_ids).ids
+                )
+
+            if not event.event_type_id and not questions_tokeep_ids:
                 event.question_ids = self._default_question_ids()
                 continue
 
@@ -242,11 +241,7 @@ class EventEvent(models.Model):
             else:
                 command = [(5, 0)]
             event.question_ids = command
-
-            # copy questions so changes in the event don't affect the event type
-            event.question_ids += event.event_type_id.question_ids.copy({
-                'event_type_id': False,
-            })
+            event.question_ids = [Command.link(question_id.id) for question_id in event.event_type_id.question_ids]
 
     @api.depends('event_slot_count', 'is_multi_slots', 'seats_max', 'registration_ids.state', 'registration_ids.active')
     def _compute_seats(self):

--- a/addons/event/models/event_question.py
+++ b/addons/event/models/event_question.py
@@ -19,19 +19,24 @@ class EventQuestion(models.Model):
         ('phone', 'Phone'),
         ('company_name', 'Company'),
     ], default='simple_choice', string="Question Type", required=True)
-    event_type_id = fields.Many2one('event.type', 'Event Type', ondelete='cascade', index='btree_not_null')
-    event_id = fields.Many2one('event.event', 'Event', ondelete='cascade', index='btree_not_null')
+    event_type_ids = fields.Many2many('event.type', string='Event Types', copy=False)
+    event_ids = fields.Many2many('event.event', string='Events', copy=False)
+    event_count = fields.Integer('# Events', compute='_compute_event_count')
     answer_ids = fields.One2many('event.question.answer', 'question_id', "Answers", copy=True)
     sequence = fields.Integer(default=10)
     once_per_order = fields.Boolean('Ask once per order',
-                                    help="If True, this question will be asked only once and its value will be propagated to every attendees."
-                                         "If not it will be asked for every attendee of a reservation.")
+                                    help="Check this for order-level questions (e.g., 'Company Name') where the answer is the same for everyone.")
     is_mandatory_answer = fields.Boolean('Mandatory Answer')
 
-    @api.constrains('event_type_id', 'event_id')
-    def _constrains_event(self):
-        if any(question.event_type_id and question.event_id for question in self):
-            raise UserError(_("Question cannot be linked to both an Event and an Event Type."))
+    @api.depends('event_ids')
+    def _compute_event_count(self):
+        event_count_per_question = dict(self.env['event.event']._read_group(
+            domain=[('question_ids', 'in', self.ids)],
+            groupby=['question_ids'],
+            aggregates=['__count']
+        ))
+        for question in self:
+            question.event_count = event_count_per_question.get(question, 0)
 
     def write(self, vals):
         """ We add a check to prevent changing the question_type of a question that already has answers.
@@ -50,16 +55,32 @@ class EventQuestion(models.Model):
         if self.env['event.registration.answer'].search_count([('question_id', 'in', self.ids)]):
             raise UserError(_('You cannot delete a question that has already been answered by attendees.'))
 
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_default_question(self):
+        if set(self.ids) & set(self.env['event.type']._default_question_ids()):
+            raise UserError(_('You cannot delete a default question.'))
+
     def action_view_question_answers(self):
         """ Allow analyzing the attendees answers to event questions in a convenient way:
-        - A graph view showing counts of each suggestions for simple_choice questions
+        - A graph view showing counts of each suggestion for simple_choice questions
           (Along with secondary pivot and list views)
         - A list view showing textual answers values for text_box questions. """
         self.ensure_one()
         action = self.env["ir.actions.actions"]._for_xml_id("event.action_event_registration_report")
-        action['domain'] = [('question_id', '=', self.id)]
+        action['context'] = {'search_default_question_id': self.id}
+        if event_id := self.env.context.get('search_default_event_id'):
+            action['context'].update(search_default_event_id=event_id)
+        # Fetch attendee answers for which the event is still linked to the question.
+        action['domain'] = [('event_id.question_ids', 'in', self.ids)]
+
         if self.question_type == 'simple_choice':
             action['views'] = [(False, 'graph'), (False, 'pivot'), (False, 'list')]
         elif self.question_type == 'text_box':
             action['views'] = [(False, 'list')]
         return action
+
+    def action_event_view(self):
+        self.ensure_one()
+        action = self.env["ir.actions.actions"]._for_xml_id("event.action_event_view")
+        action['domain'] = [('question_ids', 'in', self.ids)]
+        return action

--- a/addons/event/models/event_registration_answer.py
+++ b/addons/event/models/event_registration_answer.py
@@ -12,7 +12,7 @@ class EventRegistrationAnswer(models.Model):
 
     question_id = fields.Many2one(
         'event.question', ondelete='restrict', required=True,
-        domain="[('event_id', '=', event_id)]")
+        domain="[('event_ids', 'in', event_id)]")
     registration_id = fields.Many2one('event.registration', required=True, index=True, ondelete='cascade')
     partner_id = fields.Many2one('res.partner', related='registration_id.partner_id')
     event_id = fields.Many2one('event.event', related='registration_id.event_id')

--- a/addons/event/models/event_type.py
+++ b/addons/event/models/event_type.py
@@ -1,4 +1,4 @@
-from odoo import _, api, fields, models
+from odoo import api, fields, models
 from odoo.addons.base.models.res_partner import _tz_get
 
 
@@ -28,11 +28,11 @@ class EventType(models.Model):
                  })]
 
     def _default_question_ids(self):
-        return [
-            (0, 0, {'title': _('Name'), 'question_type': 'name', 'is_mandatory_answer': True}),
-            (0, 0, {'title': _('Email'), 'question_type': 'email', 'is_mandatory_answer': True}),
-            (0, 0, {'title': _('Phone'), 'question_type': 'phone'}),
-        ]
+        """ Get ids of default questions from ir.model.data. """
+        return self.env['ir.model.data'].sudo().search([
+            ('module', '=', 'event'),
+            ('name', 'in', ['event_question_name', 'event_question_email', 'event_question_phone'])
+        ]).mapped('res_id')
 
     name = fields.Char('Event Template', required=True, translate=True)
     note = fields.Html(string='Note')
@@ -55,8 +55,8 @@ class EventType(models.Model):
     # ticket reports
     ticket_instructions = fields.Html('Ticket Instructions', translate=True,
         help="This information will be printed on your tickets.")
-    question_ids = fields.One2many(
-        'event.question', 'event_type_id', default=_default_question_ids,
+    question_ids = fields.Many2many(
+        'event.question', default=_default_question_ids,
         string='Questions', copy=True)
 
     @api.depends('has_seats_limitation')
